packages/custom-elements/src/CustomElementInternals.ts

```diff
@@ -1,58 +1,49 @@
- * Code distributed by Google as part of the polymer project is also
- * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt The complete set of authors may be found
+ * at http://polymer.github.io/AUTHORS.txt The complete set of contributors may
+ * be found at http://polymer.github.io/CONTRIBUTORS.txt Code distributed by
+ * Google as part of the polymer project is also subject to an additional IP
+ * rights grant found at http://polymer.github.io/PATENTS.txt
  */
 
-import Native from './Patch/Native.js';
+import './Externs.js';
+import {CustomElementState as CEState} from './CustomElementState.js';
+import {CustomElementDefinition, HTMLImportElement} from './Externs.js';
+import * as Native from './Patch/Native.js';
 import * as Utilities from './Utilities.js';
-import CustomElementRegistry from './CustomElementRegistry.js';
-import CEState from './CustomElementState.js';
 
 const NS_HTML = 'http://www.w3.org/1999/xhtml';
 
 export default class CustomElementInternals {
-  /**
-   * @param {{
-   *   shadyDomFastWalk: boolean,
-   *   noDocumentConstructionObserver: boolean,
-   * }} options
-   */
-  constructor(options) {
-    /** @type {!Array<!function(!Node)>} */
-    this._patchesNode = [];
-
-    /** @type {!Array<!function(!Element)>} */
-    this._patchesElement = [];
-
-    /** @type {boolean} */
-    this._hasPatches = false;
-
-    /** @const {boolean} */
+  private readonly _patchesNode: Array<(node: Node) => void> = [];
+  private readonly _patchesElement: Array<(elem: Element) => void> = [];
+  private _hasPatches = false;
+  public readonly shadyDomFastWalk: boolean;
+  public readonly useDocumentConstructionObserver: boolean;
+
+  constructor(options: {
+    shadyDomFastWalk: boolean,
+    noDocumentConstructionObserver: boolean
+  }) {
     this.shadyDomFastWalk = options.shadyDomFastWalk;
-
-    /** @const {boolean} */
-    this.useDocumentConstructionObserver = !options.noDocumentConstructionObserver;
+    this.useDocumentConstructionObserver =
+        options.noDocumentConstructionObserver;
   }
 
-  /**
-   * @param {!Node} node
-   * @param {!function(!Element)} callback
-   * @param {!Set<!Node>=} visitedImports
-   */
-  forEachElement(node, callback, visitedImports) {
-    const sd = window['ShadyDOM'];
+  forEachElement(
+      node: Node, callback: (elem: Element) => void,
+      visitedImports?: Set<Node>) {
+    const sd = window['ShadyDom'];
     if (this.shadyDomFastWalk && sd && sd['inUse']) {
       if (node.nodeType === Node.ELEMENT_NODE) {
-        const element = /** @type {!Element} */(node);
+        const element = node as Element;
         callback(element);
       }
       // most easily gets to document, element, documentFragment
-      if (node.querySelectorAll) {
+      if ((node as Element).querySelectorAll!) {
```

Comment:
Is this `!` necessary? All `Element`s should have `querySelectorAll`.

Reply:
It is, to silence the warning that we check querySelectorAll in the if but we don't directly call it in the body of the if.